src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,320 @@
+    const float coefficient = logisticRegressor_.coefficients.at(i, 0);
+    memcpy(buffer, &coefficient, sizeof(coefficient));
+    buffer += sizeof(coefficient);
+  }
+
+  memcpy(buffer, &logisticRegressor_.intercept, sizeof(logisticRegressor_.intercept));
+  buffer += sizeof(logisticRegressor_.intercept);
+
+  return memory;
+}
+
+// mapBPEToWords takes the following arguments:
+// - the log probabilities (logProbs) of byte pair encodings (BPE)
+//   that comes from the tracebackWordScores method (which belongs to hypothesis.h in Marian)
+// - the AnnotatedText from the translated word
+// - the index of a translated sentence
+//
+// This method is responsible for mapping BPE tokens to word tokens representing them through ByteRanges.
+//
+// The words byte ranges are expected to be alphanumeric characters, and they are split using whitespaces. Moreover,
+// this method is also responsible for extracting the features that the QE model will further use. The features
+// extracted are the following:
+//
+// - meanScore = the mean of bpe's  logProbs  that a given word corresponds to
+// - minScore = the minimum bpe's logProbs that a given word corresponds to
+// - numSubWords = the number of bpe tokens that a term is made of
+// - overallMean = the mean of bpe's logProbs regarding the whole translated sentence
+//
+// The return of this function is a ByteRange vector of the words and a WordFeatures vector.
+//
+// If a translated sentence does not contain any alphanumeric character (therefore, it is made basically of the EOS
+// token), this method ignores it and returns an empty ByteRange vector of words and an empty WordFeatures vector.
+//
+// Examples:
+// Suppose that you have the following source target (A): marian is a good translation service and the translate service
+// gives you the following sentence (B):
+//
+// ma(0.15) ri(0.15) an(0.2) es(0.3) un(0.1) bu(0.3) en(0.2) ser(0.1) vi(0.2) cio(0.4) de(0.1) tra(0.4) du(0.2)
+// cción(0.1)
+//
+// The numbers that the words follow represent the logProb of each BPE token.
+//
+// Then, the result would be something like:
+// a vector where each position corresponds to the ByteRanges of the following words: marian es un buen servicio de
+// traducción. Hence, its length is 7.
+//
+// An vector of WordFeatures with length 7 where, for instance:
+//
+// the values of the first element (marian) would be:
+// - meanScores= (0.15+0.15+0.3)/3=0.2
+// - minScores= 0.15
+// - numSubWords = 3
+// - overallMean = 0.207
+//
+// the values of the second element (es) would be:
+// - meanScores= 0.3
+// - minScores= 0.3
+// - numSubWords = 1
+// - overallMean = 0.207
+//
+std::tuple<std::vector<ByteRange>, std::vector<QualityEstimator::WordFeatures>, float> QualityEstimator::mapBPEToWords(
+    const std::vector<float>& logProbs, const AnnotatedText& target, const size_t sentenceIdx) {
+  // Ignore empty target
+  if (logProbs.empty() || target.text.empty()) {
```

Comment:
`target.numWords(sentenceIdx) < 2` is the stronger condition?

Reply:
Indeed this is a stronger conditition. However, the issue w.r.t. English-Spanish model still exists. Therefore, what is your suggestion to solve English-Spanish, English-German and others models as well?

Reply:
If you mean random punctuations there's not much the MT system can do about it. It may or may not become `<unk>` in the future. Wishful thinking is have quality scores scream BAD BAD BAD BAD at inference from the model for dumb output that NMT provides. The solution should be statistical and ideally learnt because there are ambiguities.

It shouldn't be fixed in any rule based heuristic code here, if that's what you're asking.

Reply:
Commit e640262